src/main/scala/firrtl/passes/VerilogMemDelays.scala

```diff
@@ -0,0 +1,171 @@
+        ss1 ++ ss2 ++ Seq(
+          Connect(NoInfo, memPortField(mem, reader, "clk"), clk),
+          // connect latency pipes to read ports
+          Connect(NoInfo, memPortField(mem, reader, "en"), en),
+          Connect(NoInfo, memPortField(mem, reader, "addr"), addr)
+        )
+      }) ++ (s.writers flatMap {writer =>
+        // generate latency pipes for write ports (enable, mask, addr, data)
+        val clk = netlist(memPortField(s, writer, "clk"))
+        val (en, ss1) = pipe(memPortField(s, writer, "en"), s.writeLatency - 1, clk, one)
+        val (mask, ss2) = pipe(memPortField(s, writer, "mask"), s.writeLatency - 1, clk, one)
+        val (addr, ss3) = pipe(memPortField(s, writer, "addr"), s.writeLatency - 1, clk, one)
+        val (data, ss4) = pipe(memPortField(s, writer, "data"), s.writeLatency - 1, clk, one)
+        ss1 ++ ss2 ++ ss3 ++ ss4 ++ Seq(
+          Connect(NoInfo, memPortField(mem, writer, "clk"), clk),
+          // connect latency pipes to write ports
+          Connect(NoInfo, memPortField(mem, writer, "en"), en),
+          Connect(NoInfo, memPortField(mem, writer, "mask"), mask),
+          Connect(NoInfo, memPortField(mem, writer, "addr"), addr),
+          Connect(NoInfo, memPortField(mem, writer, "data"), data)
+        )
+      }) ++ (s.readwriters flatMap {readwriter =>
+        val (reader, writer) = rwMap(readwriter)
+        val clk = netlist(memPortField(s, readwriter, "clk"))
+        // generate latency pipes for readwrite ports (enable, addr, wmode, wmask, wdata)
+        val (en, ss1) = pipe(memPortField(s, readwriter, "en"), s.readLatency - 1, clk, one)
+        val (wmode, ss2) = pipe(memPortField(s, readwriter, "wmode"), s.writeLatency - 1, clk, one)
+        val (wmask, ss3) = pipe(memPortField(s, readwriter, "wmask"), s.writeLatency - 1, clk, one)
+        val (wdata, ss4) = pipe(memPortField(s, readwriter, "wdata"), s.writeLatency - 1, clk, one)
+        val (raddr, ss5) = pipe(memPortField(s, readwriter, "addr"), s.readLatency, clk, AND(en, NOT(wmode)))
+        val (waddr, ss6) = pipe(memPortField(s, readwriter, "addr"), s.writeLatency - 1, clk, one)
+        repl(memPortField(s, readwriter, "rdata")) = memPortField(mem, reader, "data")
+        ss1 ++ ss2 ++ ss3 ++ ss4 ++ ss5 ++ ss6 ++ Seq(
+          Connect(NoInfo, memPortField(mem, reader, "clk"), clk),
+          // connect latency pipes to split read ports
+          Connect(NoInfo, memPortField(mem, reader, "en"), en),
+          Connect(NoInfo, memPortField(mem, reader, "addr"), raddr),
+          // connect latency pipes to split write ports
+          Connect(NoInfo, memPortField(mem, writer, "clk"), clk),
+          Connect(NoInfo, memPortField(mem, writer, "en"), AND(en, wmode)),
+          Connect(NoInfo, memPortField(mem, writer, "mask"), wmask),
+          Connect(NoInfo, memPortField(mem, writer, "addr"), waddr),
+          Connect(NoInfo, memPortField(mem, writer, "data"), wdata)
+        )
+      })))
```

Comment:
This is pretty monolithic, there should be some variable assignments or something to break it up a little.
The flat mapping on readers, writers, and readwriters could be refactored into functions (even though they'd each only be called once, for clarity.